File: 69_Sqrt(x).java
class Solution {
    public static int mySqrt(int x) {
        if(x == 0 || x == 1) {
        	return x;
        }
        int l = 1, r = x;
        int res = 0;
        
        while(l <= r) {
        	long mid = (r - l) / 2 + l;
        	if(mid * mid == (long)x) {
        		return (int)mid;
        	}
        	if(mid * mid < (long)x){
        		l = (int)(mid + 1);
        		res = (int)mid;
        	}else {
				r = (int)(mid - 1);
			}
        }
        return res;
    }
    public static void main(String[] args) {
		System.out.println(mySqrt(8));
	}
}

File: c++/69_Sqrt(x).cpp
// sol1
class Solution {
public:
    int mySqrt(int x) {
        int l = 0, r = x, res = -1;
        
        while (l <= r) {
            int mid = l + (r-l) / 2;
            if ((long long)mid * mid <= x) {
                l += 1;
                res = mid;
            } else {
                r = mid - 1;
            }
        }
        
        return res;
    }
};
// sol2
class Solution {
public:
    int mySqrt(int x) {
        if (x == 0) {
            return 0;
        }
        
        int res = exp(0.5 * log(x));
        return (long long)(res+1)*(res+1) <= x ? res+1 : res;
    }
};